### PR TITLE
make triple parsing order independent

### PR DESCRIPTION
**Pull request by WalterBright**
It turns out the only order dependency is in `-cenv[-cppenv]`. For the rest, there's no reason for the order.
Weirdly, `+avx` uses a different scheme for order-dependency, although it is not actually order dependent.

I.e. the `-target` syntax is a bunch of hacks.

This PR does the rest in any order, and simplifies the code that does it. Some of this still needs to be redesigned, but will do that later. Since the `-target` has been totally unusable anyway, there shouldn't be any reliance on its bad behavior.

Also, unused `source` field was removed.

### CODE REVIEW

**Comment by WalterBright** [comment]
This heisenbug again:
```
macOS 12.x x64, DMD (latest)
****** FAIL debug64 std.experimental.allocator
core.exception.AssertError@std/experimental/allocator/package.d(3738): Assertion failure
```

**Review by thewilsonator — CHANGES_REQUESTED** [review]
This PR doesn't sound like a good idea, but I can't tell because there are no changes to the tests. Please update the tests for what this changes. 

If this allows vendor, arch, os, and env to be specified in anything but that order there had better be a good reason for it.

**Review by Geod24 — CHANGES_REQUESTED** [review]
I don't think this brings any benefit, and it diverges from what everyone else is doing.
CC @ibuclaw @kinke

**Comment by WalterBright** [comment]
> there are no changes to the tests

I was stymied every time I tried to write a test, because of other severe bugs in the implementation (in bugzilla), and blocker bugs in the test runner (already filed). Because of this, my plan is to bring it into working condition with a series of PRs.

> If this allows vendor, arch, os, and env to be specified in anything but that order

It does, except as noted in the comments.

> there had better be a good reason for it.

1. the ordering should serve a purpose. It does not that I can discern.
2. I can't remember the order. I had to look at the original PR every time. Your comment got the order wrong, too.
3. If a user needs to specify just a subset, why should he have to specify the previous ones?
4. The other switches don't have an order

> I don't think this brings any benefit

Aside from what I already listed above, the code is simpler and easier to follow. The remaining bumpy parts are due to more unnecessary inconsistencies in the grammar, that I plan to fix later.

> it diverges from what everyone else is doing

I was not able to find a specification for what anyone else is doing, let alone everyone. If there is one, please link to it.

To sum up, the triple implementation https://github.com/dlang/dmd/pull/12508/ was pulled and essentially abandoned 7 months ago. It did not work at all. So I have stepped in to fix it. This PR is one step in the process, and the purpose is mostly to make it easier to see what is happening by making each of the component parsers behave the same way.

**Comment by ibuclaw** [comment]
> It turns out the only order dependency is in `-cenv[-cppenv]`. For the rest, there's no reason for the order.

I've not known that to be the case, but then target strings are sanitized before the gdc build sees them, so everything relies on the explicit order `arch-vendor-os`. 

Did you try passing reordered up target strings to config.sub?

**Comment by WalterBright** [comment]
> Did you try passing reordered up target strings to config.sub?

config.sub ?

**Comment by ibuclaw** [comment]
> I was not able to find a specification for what anyone else is doing, let alone everyone. If there is one, please link to it.
> 

I'd consider config.sub the single source of truth on the matter given how ubiquitous autoconf is. If the script doesn't recognise the target triplet, then nobody is writing targets like that.

https://github.com/autotools-mirror/autoconf/blob/master/build-aux/config.sub

**Comment by WalterBright** [comment]
Yes, I manually tried reordering, it works. But putting it in the test suite? As I said, bugs in the test runner prevent cross compiling from working.

**Comment by ibuclaw** [comment]
> I don't think this brings any benefit, and it diverges from what everyone else is doing.
> CC @ibuclaw @kinke 

To be fair, we already don't do what everyone else is doing by having `+feature`, `-cenv`, and `-cppenv` encodified in the target string.

**Comment by WalterBright** [comment]
Thanks for providing a reference.

The first thing I noticed in config.sub:
```
echo Invalid configuration \`"$1"\': more than four components >&2
```
The existing triple implementation has 5.

**Comment by WalterBright** [comment]
The next thing is that the field values don't line up with the ones in `target.d`. The third thing is it's hard to find an order dependency in the code, it tries various things looking for a match: `Ambiguous whether COMPANY is present, or skipped and KERNEL-OS is two parts` and `Second component is usually, but not always the OS`. It's too bad the definition of the format is the code that implements it. Argh.

The existing implementation won't work with config.sub anyway. Also, not being order dependent does not *prevent* anyone from using an order that works with whatever other system requires.

**Comment by WalterBright** [comment]
I have a suggestion. My interest in this is to make cross compiling work. To that end, just a `-os=windows|linux|osx|freebsd` switch is needed. I will be happy to pass the baton on triple to someone who has a need for it.

**Comment by ibuclaw** [comment]
> I have a suggestion. My interest in this is to make cross compiling work. To that end, just a `-os=windows|linux|osx|freebsd` switch is needed. I will be happy to pass the baton on triple to someone who has a need for it.

I'm not sure whether it's a good idea to have two mutually exclusive options.

I don't see a reason not to allow what you are suggesting as accepted options for `-target`, but they do have to be unambiguous - which I don't think is the case for any of your examples (`-target=windows|linux|osx|freebsd` - is it 32bit or 64bit?).

**Comment by WalterBright** [comment]
> I'm not sure whether it's a good idea to have two mutually exclusive options.

We already have that for everything but os.

> `-target=windows|linux|osx|freebsd` - is it 32bit or 64bit?

Bitness unspecified will become the default, which is that of the host machine. Or can be set with -m32 or -m64.
